vendor/github.com/docker/swarmkit/manager/dispatcher/dispatcher.go

```diff
@@ -532,6 +539,21 @@ func (d *Dispatcher) register(ctx context.Context, nodeID string, description *a
 // UpdateTaskStatus updates status of task. Node should send such updates
 // on every status change of its tasks.
 func (d *Dispatcher) UpdateTaskStatus(ctx context.Context, r *api.UpdateTaskStatusRequest) (*api.UpdateTaskStatusResponse, error) {
+	// shutdownWait.Add() followed by isRunning() to ensures that
+	// if this rpc sees the dispatcher running,
+	// it will already have called Add() on the shutdownWait wait,
+	// which ensures that Stop() will wait for this rpc to complete.
+	// Note that Stop() first does Dispatcher.ctx.cancel() followed by
+	// shutdownWait.Wait() to make sure new rpc's don't start before waiting
+	// for existing ones to finish.
+	d.shutdownWait.Add(1)
```

Comment:
This looks suspect to me.
A waitgroup `Add()` is not safe to do async.
Ping @anshulpundir

Reply:
sync.WaitGroup operations should be atomic. I think this is a false alarm. I will try to follow up with the community.

FYI: https://groups.google.com/forum/#!topic/golang-nuts/Qq_h0_M51YM

Reply:
Actually it seems Add() needs to be synchronized with Wait(). Bummer.

I'll look at how to fix this.